flax/nnx/nn/recurrent.py

```diff
@@ -0,0 +1,920 @@
+# Copyright 2024 The Flax Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""RNN modules for Flax."""
+
+from typing import (
+    Any,
+    TypeVar,
+)
+from collections.abc import Callable
+from functools import partial
+from typing_extensions import Protocol
+from absl import logging
+
+import jax
+import jax.numpy as jnp
+
+from flax import nnx
+from flax.nnx import rnglib
+from flax.nnx.module import Module
+from flax.nnx.nn import initializers
+from flax.nnx.nn.linear import Linear
+from flax.nnx import sigmoid, tanh
+from flax.nnx.transforms.iteration import Carry
+from flax.typing import (
+    Dtype,
+    Initializer,
+    Shape
+)
+
+default_kernel_init = initializers.lecun_normal()
+default_bias_init = initializers.zeros_init()
+
+A = TypeVar("A")
+Array = jax.Array
+Output = Any
+
+
+class RNNCellBase(Module):
+    """RNN cell base class."""
+
+    def initialize_carry(
+        self, rngs: rnglib.Rngs | None, input_shape: tuple[int, ...]
```

Comment:
Maybe we could have `input_shape` first and make `rngs` optional.
```suggestion
        self, input_shape: tuple[int, ...], rngs: rnglib.Rngs | None = None
```